src/include/duckdb/common/single_thread_pointer.h

```diff
@@ -0,0 +1,169 @@
+#include <cstdlib>
+
+template <typename T>
+class single_thread_ptr {
+public:
+	T *ptr;              // contained pointer
+	uint32_t *ref_count; // reference counter
+
+public:
+	// Default constructor, constructs an empty single_thread_ptr.
+	constexpr single_thread_ptr() : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct empty single_thread_ptr.
+	constexpr single_thread_ptr(std::nullptr_t) : ptr(nullptr), ref_count(nullptr) {
+	}
+	// Construct a single_thread_ptr that wraps raw pointer.
+	explicit single_thread_ptr(T *p) {
+		if (p) {
+			ptr = p;
+			ref_count = new uint32_t(1);
+		}
+	}
+
+	single_thread_ptr(uint32_t *r, T *p) {
+		ptr = p;
+		ref_count = r;
+	}
+
+	template <class U>
+	single_thread_ptr(uint32_t *r, U *p) {
```

Comment:
Maybe I'm missing something but I'm not quite getting why there is a separate set of constructors with a separate template parameter here. Why is "U" used here, but "T" used elsewhere?

Reply:
The usage of two different types is for handling the conversion copy. sometimes, the left part of the assignment is the base class, while the left part is the derived class.

Reply:
I see, that makes sense. Thanks for the explanation!